### PR TITLE
[FEAT] Implement new COSR logic by recording historical Forecast Person cost windows

### PR DESCRIPTION
**Pull request by joshmcclain**
TLDR: We want to update COSR to use actual historical hourly costs, instead of average studio-wide costs. To do so, we'll start recording the effective rate for each ForecastAssignment using a new `ForecastPersonCostWindow` model.

This PR is broken down into 6 commits for easier readability. Those 6 commits do the following:
- Add the new `forecast_person_cost_windows` table and AR model
- Add a new `ForecastPersonCostWindowSyncer` utility, for updating/creating the necessary cost windows for a given `ForecastPerson` + `ForecastProject`.
- Update the Forecast daily sync cron to update the current cost windows using the above utility
- Add a migration for backfilling cost window records for all historical ForecastProjects
- Add a new `CostOfServicesRenderedCalculator` utility, for calculating the lifetime COSR metrics for a given project using its cost windows
- Update the project tracker to include the new COSR numbers, calculated via the above utility, into project snapshots

Some notes:
- I have switched this back to associating cost windows with `ForecastPerson` instead of `ForecastAssignment`.
- I am using the `notes` field on the ForecastProject model for reading the effective rate for a given subcontractor (necessary because we don't have admin users / skill trees / etc for these people). There are a few other edge cases related to contractor rates; I'll leave some inline comments indicating those.
- When I merge this code, I'll need to do so incrementally due to the multiple migrations involved. So once I have a thumbs-up on this overall approach, I plan to carve out individual PR's for each of the child commits included here.
- This PR does *not* include updating the project views to show the new COSR figures in the graphs. I'll do that as a separate PR once everything here has been merged.

### CODE REVIEW

File: lib/stacks/forecast_person_cost_window_syncer.rb [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        end_date: nil,
+        hourly_cost: hourly_cost
+      })
+    end
+  end
+
+  private
+
+  def hourly_cost
+    @_hourly_cost ||= if @forecast_person.roles.include?("Subcontractor")
```
Comment (hhff) [inline lib/stacks/forecast_person_cost_window_syncer.rb:90]: A better way to know if a `forecast_person` is a subcontractor is actually that we can't resolve an `admin_user` for them. Any reason why that can't work here?

File: lib/stacks/forecast_person_cost_window_syncer.rb [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+  end
+
+  def employee_rate
+    unless @forecast_person.admin_user.present?
+      Rails.logger.warn(
+        "No admin user associated with non-subcontractor Forecast person: "\
+        "#{@forecast_person.email}"
+      )
+
+      return 0
```
Comment (hhff) [inline lib/stacks/forecast_person_cost_window_syncer.rb:104]: This is a broken case, we should warn Stacks admins on this

File: lib/stacks/forecast_person_cost_window_syncer.rb [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    end
+
+    cost = @forecast_person
+      .admin_user
+      .approximate_cost_per_sellable_hour_before_studio_expenses_on_date(@target_date)
```
Comment (hhff) [inline lib/stacks/forecast_person_cost_window_syncer.rb:109]: This LOC is probably the performance bottleneck here. I wonder if we can a) improve it's performance, and b) multithread it (currently use the Parallel gem for that, see [here](https://github.com/sanctuarycomputer/stacks/blob/main/lib/tasks/stacks.rake#L94))

I think we have like 20 in the db pool rn

File: lib/stacks/forecast_person_cost_window_syncer.rb [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+    if cost.nil?
+      return nil
```
Comment (hhff) [inline lib/stacks/forecast_person_cost_window_syncer.rb:112]: We should notify in this case

File: lib/stacks/forecast_person_cost_window_syncer.rb [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+    end
+
+    cost.round(2)
+  end
+
+  def subcontractor_rate
+    if @forecast_project.notes.nil?
+      return hardcoded_subcontractor_rate
```
Comment (hhff) [inline lib/stacks/forecast_person_cost_window_syncer.rb:120]: Can we fallback to checking the `@forecast_person.notes` for a default rate, before falling back to `hardcoded_subcontractor_rate`?

File: lib/stacks/forecast_person_cost_window_syncer.rb [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+  end
+
+  def hardcoded_subcontractor_rate
+    historical_rate = HISTORICAL_SUBCONTRACTOR_RATES[@forecast_person.email]
+
+    if historical_rate.nil?
+      Rails.logger.warn(
+        "No rate note or historical found for subcontractor: "\
+        "#{@forecast_person.email}"
+      )
```
Comment (hhff) [inline lib/stacks/forecast_person_cost_window_syncer.rb:149]: Notify admins here too I think, also in the case of 0 (pretty sure no one has ever had a subcontractor rate of $0, ie worked for free)

File: app/models/forecast_person_cost_window.rb [hunk 7]
```diff
@@ -0,0 +1,4 @@
+class ForecastPersonCostWindow < ApplicationRecord
+  belongs_to :forecast_person, optional: true
+  belongs_to :forecast_project, optional: true
```
[the remark this answers is not in the document]
Comment (hhff) [reply app/models/forecast_person_cost_window.rb:3]: given these reference `forecast_id` and the upsert happens in a transaction, do we actually need these to be optional?

What happens in the case a person or project is deleted from Forecast?

File: app/models/project_tracker.rb [hunk 8]
```diff
@@ -195,6 +202,36 @@ def generate_snapshot!
         x: date.iso8601,
         y: acc[:accrual][:cosr_total] += cosr_for_date[:accrual]
       })
+
+      # Once we confirm that the new COSR numbers look okay in production,
+      # we can remove all of the above logic and just use the following instead.
```
Comment (hhff) [inline app/models/project_tracker.rb:207]: I'd be fine us just overwriting the old code here

File: lib/stacks/forecast_person_cost_window_syncer.rb (hunk 3 above)
Comment (hhff) [reply lib/stacks/forecast_person_cost_window_syncer.rb:109]: Probably not important rn

File: lib/stacks/forecast.rb [hunk 9]
```diff
@@ -144,4 +146,23 @@ def sync_all_assignments!
...
   end
+
+  def sync_current_cost_windows!
+    target_date = Date.today
+
+    query = ForecastAssignment.where(
+      "end_date >= ? AND start_date <= ?",
+      target_date,
+      target_date
+    ).includes(:forecast_person, :forecast_project)
```
[the remark this answers is not in the document]
Comment (hhff) [reply lib/stacks/forecast.rb:157]: FWIW, if we find the sync takes 10 minutes, I'd be fine to
- switch forecast sync to every 1 hour
- allow the sync to run against every historical record

We basically take that approach for all other things in stacks. Blow it out and recalc from the dawn of time

File: lib/stacks/forecast_person_cost_window_syncer.rb (hunk 5 above)
Comment (hhff) [reply lib/stacks/forecast_person_cost_window_syncer.rb:120]: oh ! got it ! ill think on that but this approach is likely fine

File: lib/stacks/forecast_person_cost_window_syncer.rb [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+      "person #{@forecast_person.email}, date #{@target_date}..."
+    )
+
+    if hourly_cost.nil?
+      Rails.logger.warn(
+        "No hourly cost found for Forecast project #{@forecast_project.forecast_id}, "\
+        "person #{@forecast_person.email}, date #{@target_date}; skipping..."
+      )
+
+      return
```
[the remark this answers is not in the document]
Comment (hhff) [reply lib/stacks/forecast_person_cost_window_syncer.rb:55]: cool

File: app/models/project_tracker.rb (hunk 8 above)
Comment (hhff) [reply app/models/project_tracker.rb:207]: aight

File: lib/stacks/forecast.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (hhff) [reply lib/stacks/forecast.rb:157]: > One worry I'd have is if we are still chewing through cost window regeneration and then invoicing is somehow triggered out of band.

Mmmm fair call. Given COSR is more of an internal metric that invoicing isn't dependent on, what if we uncouple the sync'ing of cost windows from the forecast sync?

Ie, run the sync'er in a "sync everything from the dawn of time" nightly? "recalc everthing" Seems like the easiest way to prevent the subtle bugs around updating historical skill trees or per-project contractor rates (we'll need to update a lot of these historically in the first few weeks after deploying, but without a "recalc everything" approach, the way the code is written it won't actually change COSR I believe).

File: lib/stacks/forecast_person_cost_window_syncer.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (hhff) [reply lib/stacks/forecast_person_cost_window_syncer.rb:55]: > Cool sounds good. With that approach we'd need to regenerate all of the old records, so maybe I should take the backfill migration and make that a maintenance task instead so that we can easily reuse it.

Still a proponent of a "recalc everything" approach to avoid anyone needing to run command line tools in response to us updating historical information

File: app/models/forecast_person_cost_window.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (hhff) [reply app/models/forecast_person_cost_window.rb:3]: > That being said, if we switch to your idea to nuke all the entries in this table and recreate with every syncing pass, this wouldn't be necessary.

👍

File: lib/stacks/forecast_person_cost_window_syncer.rb [hunk 11]
```diff
@@ -0,0 +1,156 @@
+class Stacks::ForecastPersonCostWindowSyncer
```
[the remark this answers is not in the document]
Comment (hhff) [reply lib/stacks/forecast_person_cost_window_syncer.rb:1]: > Maybe I'm misunderstanding something about the way that we incur costs for new (< 3 months?) employees. The way I see it, if someone's skill tree was wrong for three months, they've likely already been paid the wrong amount for at least 2 of those months, so we wouldn't want to go back and correct the cost windows for those previous two months, since that money has already been paid.

It's entirely possible that they conducted a skill tree, but the admin forgot to update their salary, and then quit a month later.

My point is I'd like to rule out subtle bugs, and TBH updating historical per-project contractor rates is the more likely scenario that will cause them.

> And a separate question: when a skill tree change goes into effect in the middle of a given month, is the employee's next monthly paycheck a blend of their old rate and their new rate? (For accounting purposes I think it would be cleaner if the new comp only took effect at the start of the next billing month.)

Right now a person's salary update should be effective the day their skill tree is archived.

File: lib/stacks/forecast_person_cost_window_syncer.rb [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+  HISTORICAL_SUBCONTRACTOR_RATES = {}
```
[the remark this answers is not in the document]
Comment (hhff) [reply lib/stacks/forecast_person_cost_window_syncer.rb:2]: > it would be easier for her to just send those to me in bulk and I can drop them here rather than her needing to go through every historical project in Forecast to add the rates into the notes.

Mmm I'd probably prefer we have her enter them in forecast just so that the data is at it's source of truth?

File: lib/stacks/forecast_person_cost_window_syncer.rb [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+        next
+      end
+
+      next if cost_window.forecast_project_id != @forecast_project.forecast_id
+      next if cost_window.hourly_cost == hourly_cost
+
+      max_end_date = @target_date - 1.day
+      cost_window.update!(end_date: max_end_date)
+      needs_new_cost_window = true
+    end
```
Comment (hhff) [inline lib/stacks/forecast_person_cost_window_syncer.rb:33]: Let me see if I can repeat this logic back:
- we check if the cost_window has an end_date. That means it's "closed" and there _should_ be one (with `end_date: nil` following it already (`end_date: nil` means this cost window is the current one)
- if it does, we bump forward our "running" max_end_date (as we want contiguous cost windows without gaps) incase we might need to create one in the future based on that prev cost window with that `max_end_date`
- if the cost_window is for a whole other project, ignore, we're focused on this project, onto the next one (this should probably be the first check in the each statement)
- if the hourly cost is the same, onto the next one

I feel like there's a few bugs in this code:

If a historical skill tree was changed (or a forecast project's subcontractor rate was changed), but because it's in a "closed" or existing/current cost_window, that hourly_rate won't actually be corrected.

If we're seeding cost_windows for the first time, but a forecast_person's salary has changed multiple times throughout the time they've worked on that project, it seems that we're just going to make a single window with the latest employee_rate for the whole period.

--

A suggestion for how this is written - rather than maintaining this higher level lexical scoped variable `max_end_date`, we could do something like:

```
ActiveRecord::Base.transaction do
  cost_windows.destroy_all
  (first_recorded_assignment_date..last_recorded_assignment_date).reduce({ latest_hourly_cost: nil, latest_created_window: nil }) do |acc, date|
    latest_created_window = acc[:latest_created_window]
    hourly_cost = hourly_cost_for_date(date)
    hourly_cost_did_change = hourly_cost != acc[:latest_hourly_cost]

    if hourly_cost_did_change
      latest_created_window.update(ended_at: date - 1.day)
      latest_created_window = ForecastPersonCostWindow.create!({ started_at: date, ... })
    end

    return { latest_hourly_cost: hourly_cost, latest_created_window: latest_created_window }
  end
end
```

File: lib/stacks/forecast_person_cost_window_syncer.rb [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+
+    if needs_new_cost_window
+      ForecastPersonCostWindow.create!({
+        forecast_project: @forecast_project,
+        forecast_person: @forecast_person,
+        start_date: [@target_date, max_end_date + 1.day].min,
+        end_date: nil,
+        needs_review: hourly_cost == 0,
```
[the remark this answers is not in the document]
Comment (hhff) [reply lib/stacks/forecast_person_cost_window_syncer.rb:41]: I like this approach!

File: lib/stacks/forecast_person_cost_window_syncer.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (hhff) [reply lib/stacks/forecast_person_cost_window_syncer.rb:2]: SGTM
